### ISSUE TITLE
"docker-compose down" using the version 2 file format kills weave containers

### ISSUE THREAD

**Issue by roverwolf** [weaveworks/weave#2025]
Using weave v 1.4.5, docker v1.10.2, docker-compose v1.6.2

I am not sure if this is an issue with weave or with docker-compose, but I can only reproduce the problem when using weave.

I have reproduced the case using the `docker-compose.yml` file from the weave-and-docker-platform guide's [app](https://github.com/weaveworks/guides/tree/master/weave-and-docker-platform/app) directory using a local docker instance running under Linux.

  - Change the `docker-compose.yml` file to be in the version 2 format
  - Run `docker-compose up -d`
  - Run `docker-compose down`

Following these steps will 
  - successfully start the containers on docker-compose up
  - kill the weave and weaveexec containers on docker-compose down
  - leave the started containers runing as docker can no longer communicate with the docker daemon

**Comment by bboreham** [comment]
You should be able to restore communication with Docker by running `eval $(weave env --restore)`

@roverwolf could you post the output of `docker logs weaveproxy` please?

**Comment by roverwolf** [comment]
@bboreham Yes, restoring communication with Docker is not an issue.

I posted the log to https://gist.github.com/roverwolf/a78578bd24f667ded4fa which was obtained just after the error.  A few seconds later the weaveproxy container goes away completely and does not show up in the list of `docker ps -a` (nor do the weave and weaveplugin containers).

**Comment by rade** [comment]
I followed [our guide](https://www.weave.works/guides/part-3-creating-and-scaling-multi-host-docker-deployment-with-swarm-and-compose-using-weave/), running Docker 1.10.3 and
```
$ docker-machine version
docker-machine version 0.6.0, build e27fb87
$ docker-compose version
docker-compose version 1.6.2, build 4d72027
docker-py version: 1.7.2
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1e 11 Feb 2013
```

After `./1-machine-create.sh` I checked what docker version was being used:

```
$ docker $(docker-machine config weave-1) version
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 21:49:11 2016
 OS/Arch:      linux/amd64
```

When running `./2-weave-launch.sh` I noticed that it was launching weave 1.4.1, which I confirmed with

```
$ DOCKER_CLIENT_ARGS=$(docker-machine config weave-1) ./weave version
weave script 1.4.1
weave router 1.4.1
weave proxy  1.4.1
weave plugin 1.4.1
```

I stopped weave on the three machines with
```
$ for i in $(seq 3) ; do echo "## weave-$i" ; DOCKER_CLIENT_ARGS=$(docker-machine config weave-$i) ./weave stop ; done
## weave-1
## weave-2
## weave-3
```
and then replaced the weave script with
```
curl -L https://github.com/weaveworks/weave/releases/download/v1.4.5/weave -o weave
```

I then re-ran `./2-weave-launch.sh` and confirmed the weave version with
```
$ DOCKER_CLIENT_ARGS=$(docker-machine config weave-1) ./weave version
weave script 1.4.5
weave router 1.4.5
weave proxy  1.4.5
weave plugin 1.4.5
```

I then moved on to `./3-replace-swarm-agents.sh` followed, as per the instructions, by
```
cd ../app
./build.sh
eval "$(docker-machine env --swarm weave-1)"
docker-compose up -d
docker-compose ps
```

That all worked well.

And so did this...
```
$ docker-compose down
Stopping app_redis_1 ... done
Stopping app_web_1 ... done
Removing app_redis_1 ... done
Removing app_web_1 ... done
$ docker ps
CONTAINER ID        IMAGE                        COMMAND                  CREATED             STATUS              PORTS               NAMES
44e6b0cbc7c8        weaveworks/plugin:1.4.5      "/home/weave/plugin"     27 minutes ago      Up 27 minutes                           weave-3/weaveplugin
9e4af267b3f7        weaveworks/weaveexec:1.4.5   "/home/weave/weavepro"   27 minutes ago      Up 27 minutes                           weave-3/weaveproxy
dd20d78f1af2        weaveworks/weave:1.4.5       "/home/weave/weaver -"   27 minutes ago      Up 27 minutes                           weave-3/weave
691deea33734        weaveworks/plugin:1.4.5      "/home/weave/plugin"     28 minutes ago      Up 27 minutes                           weave-2/weaveplugin
3490a8af30f5        weaveworks/weaveexec:1.4.5   "/home/weave/weavepro"   28 minutes ago      Up 27 minutes                           weave-2/weaveproxy
7716ee9a9214        weaveworks/weave:1.4.5       "/home/weave/weaver -"   28 minutes ago      Up 27 minutes                           weave-2/weave
a7a1182d5e4a        weaveworks/plugin:1.4.5      "/home/weave/plugin"     28 minutes ago      Up 28 minutes                           weave-1/weaveplugin
88aa34abd3ea        weaveworks/weaveexec:1.4.5   "/home/weave/weavepro"   28 minutes ago      Up 28 minutes                           weave-1/weaveproxy
c69aefe2fcc5        weaveworks/weave:1.4.5       "/home/weave/weaver -"   28 minutes ago      Up 28 minutes                           weave-1/weave
```
I then changed the docker-compose.yml to the v2 format:
```
$ cat docker-compose.yml
version: '2'
services:
  web:
    build: .
    command: python app.py
    ports:
       - "80:5000"
    hostname: hello.weave.local
    environment:
       - "affinity:container!=app_web_*"
  redis:
    image: redis
    hostname: redis.weave.local

$ docker-compose up -d
Creating network "app_default" with the default driver
ERROR: 500 Internal Server Error: failed to parse pool request for address space "GlobalDefault" pool "" subpool "": cannot find address space GlobalDefault (most likely the backing datastore is not configured)
```
Based on the [compose networking docs](https://docs.docker.com/compose/networking/#configuring-the-default-network) I added the following to the end of the the docker-compose.yml:
```
networks:
  default:
    driver: weavemesh
```
This allowed me to start the app...
```
$ docker-compose up -d
Creating network "app_default" with driver "weavemesh"
Creating app_web_1
Creating app_redis_1
$ docker-compose ps   Name                  Command               State              Ports            
----------------------------------------------------------------------------------
app_redis_1   /w/w /entrypoint.sh redis- ...   Up      6379/tcp                    
app_web_1     /w/w python app.py               Up      192.168.99.102:80->5000/tcp 
```
I then ran...
```
$ docker-compose down
Stopping app_redis_1 ... error
Stopping app_web_1 ... error

ERROR: for app_redis_1  Could not make request to target: unexpected EOF 
ERROR: for app_web_1  Could not make request to target: unexpected EOF 
Traceback (most recent call last):
  File "<string>", line 3, in <module>
  File "compose/cli/main.py", line 56, in main
  File "compose/cli/docopt_command.py", line 23, in sys_dispatch
  File "compose/cli/docopt_command.py", line 26, in dispatch
  File "compose/cli/main.py", line 191, in perform_command
  File "compose/cli/main.py", line 273, in down
  File "compose/project.py", line 228, in down
  File "compose/project.py", line 224, in remove_stopped
  File "compose/project.py", line 368, in containers
  File "compose/project.py", line 366, in matches_service_names
  File "compose/container.py", line 108, in labels
  File "compose/container.py", line 181, in get
  File "compose/container.py", line 230, in inspect_if_not_inspected
  File "compose/container.py", line 239, in inspect
  File ".tox/py27/lib/python2.7/site-packages/docker/utils/decorators.py", line 21, in wrapped
  File ".tox/py27/lib/python2.7/site-packages/docker/api/container.py", line 181, in inspect_container
  File ".tox/py27/lib/python2.7/site-packages/docker/client.py", line 153, in _result
  File ".tox/py27/lib/python2.7/site-packages/requests/models.py", line 812, in json
  File "json/__init__.py", line 338, in loads
  File "json/decoder.py", line 366, in decode
  File "json/decoder.py", line 384, in raw_decode
ValueError: No JSON object could be decoded
docker-compose returned -1
```
So that's not so good. Alas  `docker ps` still works just fine:
```
$ docker ps
CONTAINER ID        IMAGE                        COMMAND                  CREATED             STATUS              PORTS                         NAMES
fd099ecbc55f        redis                        "/w/w /entrypoint.sh "   3 minutes ago       Up 3 minutes        6379/tcp                      weave-3/app_redis_1
ee0a4acbe086        app_web                      "/w/w python app.py"     3 minutes ago       Up 3 minutes        192.168.99.102:80->5000/tcp   weave-3/app_web_1
44e6b0cbc7c8        weaveworks/plugin:1.4.5      "/home/weave/plugin"     38 minutes ago      Up 38 minutes                                     weave-3/weaveplugin
9e4af267b3f7        weaveworks/weaveexec:1.4.5   "/home/weave/weavepro"   38 minutes ago      Up 38 minutes                                     weave-3/weaveproxy
dd20d78f1af2        weaveworks/weave:1.4.5       "/home/weave/weaver -"   38 minutes ago      Up 38 minutes                                     weave-3/weave
691deea33734        weaveworks/plugin:1.4.5      "/home/weave/plugin"     39 minutes ago      Up 38 minutes                                     weave-2/weaveplugin
3490a8af30f5        weaveworks/weaveexec:1.4.5   "/home/weave/weavepro"   39 minutes ago      Up 39 minutes                                     weave-2/weaveproxy
7716ee9a9214        weaveworks/weave:1.4.5       "/home/weave/weaver -"   39 minutes ago      Up 39 minutes                                     weave-2/weave
a7a1182d5e4a        weaveworks/plugin:1.4.5      "/home/weave/plugin"     39 minutes ago      Up 39 minutes                                     weave-1/weaveplugin
88aa34abd3ea        weaveworks/weaveexec:1.4.5   "/home/weave/weavepro"   39 minutes ago      Up 39 minutes                                     weave-1/weaveproxy
c69aefe2fcc5        weaveworks/weave:1.4.5       "/home/weave/weaver -"   39 minutes ago      Up 39 minutes                                     weave-1/weave
```

But for some reason swarm is having trouble talking to weave-3:
```
$ docker logs weave-1/weaveplugin
INFO: 2016/03/20 16:37:51.945993 Weave plugin 1.4.5 Command line options: []
INFO: 2016/03/20 16:37:51.946875 Docker API on unix:///var/run/docker.sock: &[Os=linux Arch=amd64 KernelVersion=4.1.19-boot2docker BuildTime=2016-03-10T21:49:11.235199091+00:00 Version=1.10.3 ApiVersion=1.22 GitCommit=20f81dd GoVersion=go1.5.3]
INFO: 2016/03/20 16:37:51.946973 Listening on /run/docker/plugins/weave.sock for global scope
INFO: 2016/03/20 16:37:51.947041 Listening on /run/docker/plugins/weavemesh.sock for local scope
INFO: 2016/03/20 16:37:51.964901 Create network 25856177243a187bbce279886ac5b45a02460a2068245d45177e0928fe0db62e
$ docker logs weave-3/weaveplugin
invalid character 'C' looking for beginning of value
```
and so does machine...
```
$ docker $(docker-machine config weave-3) logs weaveplugin
Error running connection boilerplate: Error checking and/or regenerating the certs: There was an error validating certificates for host "192.168.99.102:2376": dial tcp 192.168.99.102:2376: getsockopt: connection refused
You can attempt to regenerate them using 'docker-machine regenerate-certs [name]'.
Be advised that this will trigger a Docker daemon restart which will stop running containers.

Error: No such container: weaveplugin
```

but the machine is alive and I can ssh into it with `docker-machine ssh weave-3`. Looking at the docker logs in `/var/log/docker.log` shows that the docker daemon died:

```
time="2016-03-20T17:16:27.482365463Z" level=debug msg="GET /v1.15/containers/json?all=1&size=0&filters={\"id\":[\"fd099ecbc55f6a33a2352c68196d6d05bcb573944c9f69723ae5c23ee99a1710\"]}" 
time="2016-03-20T17:16:27.491800040Z" level=debug msg="Remove interface vethwgfa30c failed: Link not found" 
panic: runtime error: index out of range

goroutine 5355 [running]:
github.com/docker/libnetwork/etchosts.Delete(0xc82078ecb0, 0x6a, 0xc821906e00, 0x5, 0x8, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/etchosts/etchosts.go:154 +0x543
github.com/docker/libnetwork.(*sandbox).deleteHostsEntries(0xc821545180, 0xc821906e00, 0x5, 0x8)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/sandbox.go:815 +0x6c
github.com/docker/libnetwork.(*endpoint).sbLeave(0xc821906900, 0x7f9152764fb8, 0xc821545180, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/endpoint.go:599 +0xb35
github.com/docker/libnetwork.(*endpoint).Leave(0xc8213a2500, 0x7f9152764fb8, 0xc821545180, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/endpoint.go:536 +0x1bf
github.com/docker/libnetwork.(*sandbox).delete(0xc821545180, 0xc82122be00, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/sandbox.go:199 +0xad4
github.com/docker/libnetwork.(*sandbox).Delete(0xc821545180, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/sandbox.go:166 +0x32
github.com/docker/docker/daemon.(*Daemon).releaseNetwork(0xc820001c80, 0xc821544e00)
	/go/src/github.com/docker/docker/daemon/container_operations_unix.go:1013 +0x4f1
github.com/docker/docker/daemon.(*Daemon).Cleanup(0xc820001c80, 0xc821544e00)
	/go/src/github.com/docker/docker/daemon/start.go:159 +0x3d
github.com/docker/docker/container.(*containerMonitor).Close(0xc820b87e00, 0x0, 0x0)
	/go/src/github.com/docker/docker/container/monitor.go:127 +0x68
github.com/docker/docker/container.(*containerMonitor).start.func1(0xc8212e9d5f, 0xc820b87e00, 0xc8212e9d60)
	/go/src/github.com/docker/docker/container/monitor.go:158 +0xaa
github.com/docker/docker/container.(*containerMonitor).start(0xc820b87e00, 0x0, 0x0)
	/go/src/github.com/docker/docker/container/monitor.go:239 +0xbc5
github.com/docker/docker/container.(*containerMonitor).(github.com/docker/docker/container.start)-fm(0x0, 0x0)
	/go/src/github.com/docker/docker/container/monitor.go:101 +0x2e
github.com/docker/docker/pkg/promise.Go.func1(0xc820fbfaa0, 0xc82087d0a0)
	/go/src/github.com/docker/docker/pkg/promise/promise.go:8 +0x1d
created by github.com/docker/docker/pkg/promise.Go
	/go/src/github.com/docker/docker/pkg/promise/promise.go:9 +0x5d
...
```

That is caused by a [docker bug](https://github.com/docker/libnetwork/issues/966).

If that bug had hit weave-1 instead of weave-3, then swarm would have broken completely, which would result in the symptoms the OP saw.

It is a pity that the fix for the bug was not included in Docker 1.10.3, so I guess we will all have to wait until 1.11.
